### ISSUE TITLE
[Attention] 超级视客营 MMCV

### ISSUE THREAD

**Issue by HAOCHENYE** [open-mmlab/mmcv#2421]
![image](https://user-images.githubusercontent.com/57566630/202351166-b68a3161-a363-4403-8e50-f699a19466d1.png)
# 活动介绍

大家好，第一期 OpenMMLab 超级视客营实训活动开始啦！超级视客营实训活动提供十七个方向、上百个不同难度的任务供大家选择，不管你是初涉 AI 的新手还是资深炼丹师，都有适合你的任务供你选择。助力大家上手 OpenMMLab 开源算法库并参与项目建设。本期活动联合北京超级云计算中心，提供算力支持，为大家开发保驾护航。

活动参与方式：选择你感兴趣的任务，在 OpenMMLab 官网提交报名表。完成匹配后，即可和导师对接制定任务规划，开始上手开发。根据不同任务要求在对应的地址提交代码结果，出题方初步 review 通过后即可领取下一个任务或者坐等领奖。活动详情戳：[OpenMMLab 官网活动页](https://openmmlab.com/activity/codecamp)。

# Task description

随着MMCV支持越来越多的下游存储库，越来越多的算子被添加到 MMCV 中。然而，并不是所有的算子都适应所有的硬件，例如，我们只提供了 `softnms` 的CPU实现，而尚未 GPU 实现，`focal loss` 只提供了 GPU 实现而没有提供 CPU 实现。

为了让大家能够更容易地在不同的设备上运行 MMCV，我们呼吁大家一起参与到算子实现的活动中来，通过这次活动，你将从中学习：

1. 规范的 PR 流程
2. 熟悉向 MMCV 添加算子的流程
3. 掌握如何在 C++/CUDA 层面扩展 PyTorch

**如果你对此很感兴趣，可以从下表中认领任务 [here](https://openmmlab.com/activity/codecamp/apply)**

提交 PR 的完整流程可以参考 [guide](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md).

| 算子名称 | 支持硬件 | 难度 | 积分 | 
| :-: | :-: | :-: | :-: |
| [softnms](#softnms) | CUDA | ⭐️⭐️⭐️ | 30  |
| [nms_match](#nms_match) | CUDA | ⭐️⭐️⭐️ | 30 |
| [sigmoid_focal_loss](#sigmoid_focal_loss) | CPU | ⭐️⭐️⭐️ | 30  |
| [softmax_focal_loss](#softmax_focal_loss) | CPU | ⭐️⭐️⭐️ | 30 |
| [assign_score_withk](#assign_score_withk) | CPU | ⭐️⭐️⭐️ | 30|
| [bbox_overlaps](#bbox_overlaps) | CPU | ⭐️⭐️⭐️ | 30 |
| [correlation](#correlation) | CPU | ⭐️⭐️⭐️⭐️ | 80 |
| [ms_deform_attn](#ms_deform_attn) | CPU | ⭐️⭐️⭐️⭐️⭐️ | 80 |

# 模板 PR

PR [2365](#2365) 提供了 [softnms](#softnms), [sigmoid_focal_loss](#sigmoid_focal_loss) and [softmax_focal_loss](#softmax_focal_loss) 三个算子的任务模板. 它只完成了算子的注册、分发、以及单元测试部分的工作，你仍然需要完成算子的具体实现，目前各个算子只会打印 "Please implement xxx"。

# Operators

<h2 id="softnms">softnms</h3>

- CPU 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/pytorch/cpu/nms.cpp#L59

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_nms.py#L41

- 技术标签
   C++; CUDA; Python; Detection

<h2 id="nms_match">nms_match</h3>

- CPU 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/pytorch/cpu/nms.cpp#L168

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_nms.py#L110

- 技术标签
   C++; CUDA; Python; Detection

<h2 id="sigmoid_focal_loss">sigmoid_focal_loss</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/sigmoid_focal_loss_cuda_kernel.cuh#L12

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_focal_loss.py#L63

- 技术标签
   C++; CUDA; Python

<h2 id="softmax_focal_loss">softmax_focal_loss</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/softmax_focal_loss_cuda_kernel.cuh#L12

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_focal_loss.py#L42

- 技术标签
   C++; CUDA; Python

<h2 id="assign_score_withk">assign_score_withk</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/assign_score_withk_cuda_kernel.cuh#L20

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_assign_score_withk.py#L10

- 技术标签
   C++; CUDA; Python

<h2 id="bbox_overlaps">bbox_overlaps</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/bbox_overlaps_cuda_kernel.cuh#L33

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_bbox.py#L9

- 技术标签
   C++; CUDA; Python; Detection

<h2 id="correlation">correlation</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/correlation_cuda.cuh#L36

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_correlation.py#L19

- 技术标签
   C++; CUDA; Python; Detection

<h2 id="ms_deform_attn">ms_deform_attn</h3>

- CUDA 参考实现:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/mmcv/ops/csrc/common/cuda/ms_deform_attn_cuda_kernel.cuh#L18

- 需要更新的单元测试:

    https://github.com/open-mmlab/mmcv/blob/652b1bf207616608179398abdcf17276b4a0ab27/tests/test_ops/test_ms_deformable_attn.py#L107

- 技术标签
   C++; CUDA; Python; Detection

活动报名地址：[报名表地址](https://openmmlab.com/activity/codecamp/apply)
根据任务难度可以获得对应积分，兑换不同奖品。另外完成任务后在知识社区发布学习心得即可获得额外积分（记得主动找小助手领取哦）。
活动交流群：群二维码

![image](https://user-images.githubusercontent.com/57566630/202354208-bf9646f8-0f79-4a7f-b083-f7ff94e0da39.png)

有任何疑问欢迎大家加入群聊或者 Issue 下参与讨论，快来完成挑战，加入 [OpenMMLab 贡献者队伍](https://openmmlab.com/contributor/path)吧～
